exec/Miner.hs

```diff
@@ -331,80 +197,95 @@ getWork = do
         a = envArgs e
         m = envMgr e
 
--- | A supervisor thread that listens for new work and manages mining threads.
---
-mining :: (TargetBytes -> HeaderBytes -> RIO Env HeaderBytes) -> WorkBytes -> RIO Env ()
-mining go wb = do
-    race updateSignal (go tbytes hbytes) >>= traverse_ miningSuccess
-    getWork >>= traverse_ (mining go)
-  where
-    T3 (ChainBytes cbs) tbytes hbytes@(HeaderBytes hbs) = unWorkBytes wb
+-- -------------------------------------------------------------------------- --
+-- Mining
 
-    chain :: IO Int
-    chain = chainIdInt <$> runGet decodeChainId cbs
+workKey :: WorkBytes -> RIO Env UpdateKey
+workKey wb = do
+    e <- ask
+    T2 url _ <- NEL.head <$> readIORef (envUrls e)
+    cid <- liftIO $ chain cbytes
+    return $! UpdateKey
+        { _updateKeyChainId = cid
+        , _updateKeyHost = baseUrlHost url
+        , _updateKeyPort = baseUrlPort url
+        }
+  where
+    T3 cbytes _ _ = unWorkBytes wb
 
-    height :: IO Word64
-    height = _height <$> runGet decodeBlockHeight (B.take 8 $ B.drop 258 hbs)
+-- | A supervisor thread that listens for new work and manages mining threads.
+--
+-- TODO: use exponential backoff instead of fixed delay when retrying.
+-- (restart retry sequence when the maximum retry count it reached)
+--
+mining :: (TargetBytes -> HeaderBytes -> RIO Env HeaderBytes) -> RIO Env ()
+mining go = do
+    updateMap <- newUpdateMap -- TODO use a bracket
+    miningLoop updateMap go
 
-    -- TODO Rework to use Servant's streaming? Otherwise I can't use the
-    -- convenient client function here.
-    updateSignal :: RIO Env ()
-    updateSignal = catchAny f $ \se -> do
-        logWarn "Couldn't connect to update stream. Trying again..."
-        logDebug $ display se
-        threadDelay 1000000  -- One second
-        updateSignal
+-- | The inner mining loop.
+--
+-- It uses 'getWork' to obtain new work and starts mining. If a block is solved it
+-- calls 'miningSuccess' and loops around.
+--
+-- It also starts over with new work when an update is triggered.
+--
+-- NOTE: if this fails continuously, e.g. because of a miss-configured or buggy
+-- miner, this function will spin forever!
+--
+-- TODO: add rate limiting for failures and raise an error if the function fails
+-- at an higher rate.
+--
+miningLoop :: UpdateMap -> (TargetBytes -> HeaderBytes -> RIO Env HeaderBytes) -> RIO Env ()
+miningLoop updateMap inner = mask $ \umask ->
+    void (go umask) `finally` logInfo "Mining halted."
+  where
+    go :: (RIO Env () -> RIO Env a) -> RIO Env b
+    go umask = do
+        forever (umask loopBody) `catchAny` \e -> do
+            logWarn "Error in mining loop. Trying again ..."
+            logDebug $ display e
+        go umask
```

Comment:
Should this truly loop forever?

Reply:
I think, it's fine for now. But, generally, it should discriminate on the type of failure. I think, there are three relevant classes of failures:

1. permanent failure, from which the loop wouldn't recover. Stop mining  loop
2. temporary failure that would require exponential backoff to recover
3. (uniform) random failure. Restart immediately but rate limiting (with burst buffer and upper failure limit) might be good.